### PR TITLE
Remove extra bos for prompt/response data with llama3.1

### PR DESCRIPTION
**Pull request by dakinggg**
This is a modification to https://github.com/mosaicml/llm-foundry/pull/1003 and the explanation from there still stands. The difference is that llama3.1 does not have an `add_bos_token` attr because it is not a `LlamaTokenizerFast`, but is a `PreTrainedTokenizerFast`.

Existing unit test covers the existing usage, and manual test for llama3.1 is here:
```
In [1]: from llmfoundry.data.finetuning.tasks import _tokenize_with_bos_removal

In [2]: import transformers

In [3]: llama318t = transformers.AutoTokenizer.from_pretrained('meta-llama/Meta-Llama-3.1-8B-Instruct')

In [4]: llama318t(text='hello', text_target='goodbye')
Out[4]: {'input_ids': [128000, 15339], 'attention_mask': [1, 1], 'labels': [128000, 19045, 29474]}

In [5]: _tokenize_with_bos_removal(llama318t, text='hello', text_target='goodbye')
Out[5]: {'input_ids': [128000, 15339], 'attention_mask': [1, 1], 'labels': [19045, 29474]}
```

Before and after loss curve. Similarly to the previous PR, loss is lower after because we're not forcing the model into an unfamiliar use of bos.
Screenshot 2024-08-21 at 6 29 21 PM

### CODE REVIEW

**Review by snarayan21 — APPROVED** [review]
lgtm! makes sense that this impacts loss, still wild to see though
